agreement/player.go

```diff
@@ -398,20 +398,17 @@ func (p *player) partitionPolicy(r routerHandle) (actions []action) {
 		actions = append(actions, a0)
 	}
 
-	/*
-	 * Important Comment:
-	 * The spec, in particular, says that during resynchronization, we must relay
-	 * the pinned payload on resynchronization. In addition, on receiving a next bundle,
-	 * we must set the pinned value.
-	 * The implementation currently does not follow the spec, and instead relies on the
-	 * Reproposal Payload mechanism (where players relay payloads on seeing reproposals)
-	 * to synchronize knowledge of Payloads across the network. In particular, this seems
-	 * to violate a termination invariant under common-case synchrony, since relaying Payloads
-	 * is now mildly delayed and we can enter periods in this strange case.
-	 * - There is significant conversation to be had around here, and pinned values,
-	 * to decide on the best course of action. For now, I will leave the implementation as-is
-	 * and defer to this future conversation as a todo.
-	 */
+	// On resynchronization, first try relaying the staged proposal from the same period as
+	// the freshest bundle. If that does not exist, for instance if we saw two next quorums in a row,
+	// then we fall back to relaying the pinned value, for liveness.
+	// One specific scenario where this is essential, assuming we handle ensure digest asynchronously:
+	// - Let the majority of honest nodes cert vote, and then see a next value quorum, and enter p + 1.
+	// - They see another next value quorum, and enter p + 2.
+	// - The minority of honest nodes see a certThreshold (but without a block), in period p. Assume that
+	//   they are partitioned from the majority of honest nodes, until the majority reach p + 2.
+	// - The minority already has the freshest bundle, so will not advance to period p + 2. However, the
+	//   majority will also filter out the cert threshold (due to a stale period). The only way to resolve
```

Comment:
I think this is an unrelated bug, but perhaps we should not filter out cert thresholds from stale periods. Perhaps we should use the same logic for determining the freshest bundle as we do in `bundleFresh` (inside of `voteAggregator`); that way, we precisely filter out the bundles that we know to be stale. (There is no change in denial-of-service characteristics here, as it is always trivial to generate a bundle with a malformed vote for a next period, and that bundle will always be fresh.) It might be possible to handle bundle filtering at the `voteTrackerRound`, since that's the machine that knows about the freshest bundle.

This is unrelated to this bug so probably the thing to do is to make an issue for it.

Reply:
I will probably rope this into my ensure digest PR

Reply:
Sounds good to me.